### ISSUE TITLE
Add RaggedTensor support to tf.stop_gradient and tf.reduce_logsumexp

### ISSUE THREAD

**Issue by ivodopyanov** [tensorflow/tensorflow#45569]
**System information**
- TensorFlow version (you are using): 2.4.0rc3
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.** Tensorflow prohibits use of tf.stop_gradient or tf.reduce_logsumexp with RaggedTensor. Other tf.math.reduce_XXX methods support RaggedTensor though.

**Will this change the current api? How?** Should not.

**Who will benefit with this feature?** In my case I was implementing gumbel-softmax with straight-through estimator for seq2seq with sampling from multiple positions in the same sequence. Straight-through estimator requires tf.stop_gradient call. Doing this with RaggedTensor would've been much more efficient.

**Any Other info.**

**Comment by edloper** [comment]
I will see what we can do about adding these.  In the meantime, you could use something like this to stop gradients on ragged tensors:

```python
def stop_ragged_gradient(rt):
  return rt.with_flat_values(tf.stop_gradient(rt.flat_values))
```

And for logexpsum, you could use something along the lines of:

```python
def reduce_logsumexp(input_tensor, axis=None, keepdims=False):
  raw_max = tf.reduce_max(input_tensor, axis, keepdims=True)
  raw_max = tf.where(tf.math.is_finite(raw_max), raw_max, tf.zeros([], raw_max.dtype))
  if isinstance(raw_max, tf.RaggedTensor):
    raw_max = raw_max.with_flat_values(tf.stop_gradient(raw_max.flat_values))
  else:
    raw_max = tf.stop_gradient(raw_max)
  result = raw_max + tf.math.log(
      tf.reduce_sum(tf.math.exp(input_tensor - raw_max),
                    axis=axis, keepdims=True))
  if not keepdims:
    result = tf.squeeze(result, axis)
  return result
```

(c.f. the source code for tf.reduce_logsumexp [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L3122).)

**Comment by ivodopyanov** [comment]
Thanks. I also found another way suitable for me - using common tf.boolean_mask and tf.scatter_nd.

**Comment by rmothukuru** [comment]
@ivodopyanov,
Can you please confirm if we can close this issue. Thanks!

**Comment by foxik** [comment]
@rmothukuru I think supporting RaggedTensors across the board is a good idea anyway (according to https://www.tensorflow.org/guide/ragged_tensor, they are supported "by more than a hundred TF ops"), so I would not close the issue just because there is a workaround.

**Comment by mcmar** [comment]
Since making `tf.stop_gradient` work with ragged tensors appears to be trivial to do, but nonetheless presents a footgun to the end user, can we prioritize adding this functionality to `tf.stop_gradient`? I'm fine for now using this workaround, but it'd be nice if new users didn't have to google `"tf stop_gradient ragged tensor"` and find this issue, then copy-paste a function to make this work.
